### ISSUE TITLE
2021/11/30-2021/12/09

### ISSUE THREAD

**Issue by njs03332** [njs03332/nlp_study#12]
- 각자 목표 진도를 정하자!!!
- 단비님: 3장 끝까지
- 선미님: 4.4, 가능하면 5.2까지
- 유리님: 4.2, 4.4장
- 다음주 스터디: 12/9 목요일 9:00PM

**Comment by njs03332** [comment]
### 3.6. 예제: 레스토랑 리뷰 감성 분류하기
#### 3.6.3 Vocabulary, Vectorizer, DataLoader 클래스
91쪽.
- 원 핫 벡터의 제약: 희소한 배열, 단어 순서를 무시한다
- 다음 장에서 이러한 제약을 줄이는 다른 방법 소개 -> 이게 뭘까? (다음 장 공부할 때 기억하기 위해 기록..)
#### 3.6.4 퍼셉트론 분류기
- pyTorch `nn.Module`에 대하여 (추가 조사)
  - `forward` 함수만 정의하고 나면, (변화도를 계산하는) `backward` 함수는 autograd 를 사용하여 자동으로 정의됨
  - 모델의 학습 가능한 매개변수들은 `.parameters()` 에 의해 반환됨
  - `forward()` 함수는 객체를 데이터와 함께 호출하면 자동으로 실행됨 
    - 예) model이란 이름의 객체를 생성 후, `model(입력 데이터)`와 같은 형식으로 객체를 호출하면 자동으로 forward 연산이 수행됨
  - `state_dict()` : In PyTorch, the learnable parameters (i.e. weights and biases) of a torch.nn.Module model are contained in the model’s parameters (accessed with model.parameters()). A state_dict is simply a Python dictionary object that maps each layer to its parameter tensor.

참고자료: https://wikidocs.net/60036, https://tutorials.pytorch.kr/beginner/blitz/neural_networks_tutorial.html, https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html

#### 3.6.5 모델 훈련
- 모델 저장: `torch.save`
  - https://jdjin3000.tistory.com/17
- `args`의 `early_stopping_criteria` = patience에 해당?
  - patience: 개선이 안된다고 바로 종료시키지 않고, 개선을 위해 몇번의 에포크를 기다릴지 설정하는 값
  - 예) 5번 이상 연속으로 손실이 나빠지면 훈련을 조기종료함
  - `update_train_state` 함수에서 `if loss_t >= train_state['early_stopping_best_val']:`
    - '손실이 나빠지면' 이 아니라 '손실이 특정값(`train_state['early_stopping_best_val']`)보다 크면' 이 아닌가?
- `loss_t = loss.item()`
  - `t.item()` for a tensor t simply converts it to python's default float32
  - we use `t.item()` to maintain running loss instead of t because PyTorch tensors store history of its values which might overload your GPU very soon
  - https://stackoverflow.com/questions/61092523/what-is-running-loss-in-pytorch-and-how-is-it-calculated
- Learning rate scheduler
  - 학습이 진행되면서 학습률을 그 상황에 맞게 가변적으로 적당하게 변경될 수 있다면 더 낮은 손실값을 얻을 수 있습니다. 이를 위해 학습률을 스케쥴이 필요합니다.
  - 예) `ReduceLROnPlateau`: 원하는 에폭마다, 이전 학습률 대비 변경폭에 따라 학습률을 감소시켜주는 방식
  - http://www.gisdeveloper.co.kr/?p=8443

**Comment by givitallugot** [comment]
#### @classmethod, @staticmethod
@classmethod - load_dataset_and_load_vectorizer: 새로운 객체 생성 시
@staticmethod - load_dataset_and_load_vectorizer: 객체 재사용 시 

python의 3가지 method
1. static method: 인스턴스화 하지 않고 메소드를 호출 할 수 있음. 독립적으로 사용되기 때문에 다른 속성에 액세스하거나 해당 클래스 내의 다른 메소드를 호출 할 수 없음.
2. class method: 인스턴스화 하지 않고 메소드를 호출 할 수 있음.  다른 메소드 및 클래스 속성에 액세스 할 수 있는 기능에 의존하지만 인스턴스 속성은 없음.
3. instance: 인스턴스화 했을때만 호출이 가능. self를 통해 해당 클래스의 모든 속성에 액세스 할 수 있음.

static method과 class method을 굳이 나누는 이유는? 
- static method은 객체 필드와 독립적이지만 로직상 클래스내에 포함되는 메서드에 사용
- 보통 클래스 속성에 엑세스할 필요가 있을 때는 class method를, 이를 엑세스할 필요가 없을 때는 static method를 사용
<br>

#### cls, self 차이
메소드 타입의 차이. instance 매소드를 사용하면 self 키워드를 사용하고 class 매소드를 사용하면 cls 키워드를 사용
static 매소드는 속성에 접근할 수 없기 때문에 사용하는 키워드가 없음.

```python
class Cs:
    
    @staticmethod
    def static_method():
        print("Static method")
 
    @classmethod
    def class_method(cls):
        print("Class method")
 
    def instance_method(self):
        print("Instance method")
 
i = Cs()
Cs.static_method()
Cs.class_method()
i.instance_method() 
```
<br>

#### Python Convention
_(언더스코어)와 __(더블언더스코어) 차이

- 둘 다 클래스 내에서 private하게 사용할 것이니 외부에서 접근하지 말라는 의미
- 그러나 _(언더스코어)는 접근 제한을 권유하되 강제하진 않음 - semiprivate
- __(더블언더스코어)는 접근 자체를 강제함 - superprivate

```python
class MyClass():
	def __init__(self):
		self._semiprivate = “Hello”
		self.__superprivate = “,, world!”
— 
	
from MyClass import *
mc = MyClass()
print(mc._semiprivate) # “Hello”
print(mc.__superprivate) # Error
```

**Comment by danbi5228** [comment]
### string.punctuation
string.punctuation; 구두점 모음 string

```python
# command
print(string.punctuation, '\ntype: ', type(string.punctuation))

# output
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ 
type:  <class 'str'>
```

---

### unpacking **

```python3
dict_ = {'a': 1, 'b': 2}

def function_return_json(d):
    return function_print(**d)

def function_print(a, b):
    print(a, b)
    
function_return_json(dict_)

# output
1 2
```

---

### 질문
- p.94 for 구문의 dataloader 형태..?
-
